include/samurai/petsc/block_assembly.hpp

```diff
@@ -0,0 +1,132 @@
+            std::size_t i = 0;
+            for_each(_operators, [&](auto& op)
+            {
+                auto row = i / cols;
+                auto col = i % cols;
+                if (row == col)
+                {
+                    names[col] = op.unknown().name();
+                }
+                i++;
+            });
+            return names;
+        }
+
+        void create_matrix(Mat& A)
+        {
+            std::size_t i = 0;
+            for_each(_operators, [&](auto& op)
+            {
+                /*auto row = i / cols;
+                auto col = i % cols;
+                std::cout << "create_matrix (" << row << ", " << col << ")" << std::endl;*/
+                op.create_matrix(_blocks[i]);
+                i++;
+            });
+
+            MatCreateNest(PETSC_COMM_SELF, rows, PETSC_NULL, cols, PETSC_NULL, _blocks.data(), &A);
+        }
+
+        void assemble_matrix(Mat& A)
+        {
+            std::size_t i = 0;
+            for_each(_operators, [&](auto& op)
+            {
+                /*auto row = i / cols;
+                auto col = i % cols;
+                std::cout << "assemble_matrix (" << row << ", " << col << ")" << std::endl;*/
+                op.assemble_matrix(_blocks[i]);
+                i++;
+            });
+            MatAssemblyBegin(A, MAT_FINAL_ASSEMBLY);
+            MatAssemblyEnd(A, MAT_FINAL_ASSEMBLY);
+        }
+
+        void enforce_bc(std::array<Vec, rows>& b) const
+        {
+            std::size_t i = 0;
+            for_each(_operators, [&](const auto& op) 
+            {
+                auto row = i / cols;
+                //auto col = i % cols;
+                if (op.include_bc())
+                {
+                    //std::cout << "enforce_bc (" << row << ", " << col << ") on b[" << row << "]" << std::endl;
+                    op.enforce_bc(b[row]);
+                }
+                i++;
+            });
+        }
+
+        std::array<Vec, cols> create_solution_vectors() const
+        {
+            std::array<Vec, cols> x_blocks;
+            std::size_t i = 0;
+            for_each(_operators, [&](const auto& op) 
+            {
+                auto row = i / cols;
+                auto col = i % cols;
+                if (row == 0)
+                {
+                    x_blocks[col] = create_petsc_vector_from(op.unknown());
+                    PetscObjectSetName(reinterpret_cast<PetscObject>(x_blocks[col]), op.unknown().name().c_str());
+                }
+                i++;
+            });
+            return x_blocks;
+        }
+
+        template<class... Fields>
+        auto tie(Fields&... fields) const
+        {
+            static constexpr std::size_t n_fields = sizeof...(fields);
+            static_assert(n_fields == cols, "The number of fields must correspond to the number of columns of the block operator.");
+
+            return std::tuple<Fields&...>(fields...);
+        }
```

Comment:
This method is used to facilitate the creation of the right-hand side of the system:
```
auto rhs = stokes.tie(f, zero);
block_solver.solve(rhs);
```
It also checks the number of fields because we can.
This check is already done in the solve() method.

Now I have added a solve() method with variadic arguments:
`block_solver.solve(f, zero);`
so no need to use tie() anymore.